catalog/app/utils/json-schema/json-schema.ts

```diff
@@ -0,0 +1,221 @@
+
+  if (!Array.isArray(prop(condition, optSchema))) return false
+
+  return (optSchema[condition] as JsonSchema[])
+    .filter(R.has('type'))
+    .some((subSchema) => doesTypeMatchSchema(value, subSchema))
+}
+
+// Purpose is to find mismatch explicitly
+// TODO: rename and redesign function to avoid "if no schema -> return true aka 'type matches schema'"
+export function doesTypeMatchSchema(value: any, optSchema?: JsonSchema): boolean {
+  if (!optSchema) return true
+  return R.cond<JsonSchema, boolean>([
+    [isSchemaEnum, () => R.includes(value, R.propOr([], 'enum', optSchema))],
+    [
+      (s) => Array.isArray(prop('type', s)),
+      () =>
+        (optSchema.type as JsonSchema[]).some((subSchema) =>
+          doesTypeMatchSchema(value, subSchema),
+        ),
+    ],
+    [isSchemaAnyOf, () => doesTypeMatchCompoundSchema(value, 'anyOf', optSchema)],
+    [isSchemaOneOf, () => doesTypeMatchCompoundSchema(value, 'oneOf', optSchema)],
+    [isSchemaAllOf, () => doesTypeMatchCompoundSchema(value, 'allOf', optSchema)],
+    [isSchemaArray, () => Array.isArray(value)],
+    [isSchemaObject, () => R.is(Object, value)],
+    [isSchemaString, () => R.is(String, value)],
+    [isSchemaInteger, () => Number.isInteger(value)],
+    [isSchemaNumber, () => R.is(Number, value)],
+    [isSchemaNull, () => R.equals(null, value)],
+    [isSchemaBoolean, () => R.is(Boolean, value)],
+    [R.T, R.T], // It's not a user's fault that we can't handle the type
+  ])(optSchema)
+}
+
+export const EMPTY_SCHEMA = {}
+
+export function makeSchemaValidator(optSchema?: JsonSchema) {
+  const schema = optSchema || EMPTY_SCHEMA
+
+  const ajv = new Ajv({ useDefaults: true, schemaId: 'auto' })
+
+  try {
+    const validate = ajv.compile(schema)
+
+    return (obj: any) => {
+      validate(R.clone(obj))
+      // TODO: add custom errors
+      return validate.errors || []
+    }
+  } catch (e) {
+    // TODO: add custom errors
+    return () => [e]
+  }
+}
+
+function scanSchemaAndPrefillValues(
+  callback: (v?: any, s?: JsonSchema) => any,
+  optValue: Record<string, any>,
+  optSchema?: JsonSchema,
+): Record<string, any> {
+  if (!optSchema) return optValue
+
+  if (!optSchema?.properties) return optValue
+
+  return Object.keys(optSchema.properties).reduce((memo, key) => {
+    const valueItem = optValue === undefined ? undefined : optValue[key]
+
+    // don't touch user's primitive value
+    if (valueItem && !R.is(Object, valueItem)) return memo
+
+    const schemaItem = R.propOr({}, key, optSchema.properties) as JsonSchema
+
+    if (schemaItem.properties)
+      return R.assoc(
+        key,
+        scanSchemaAndPrefillValues(callback, valueItem, schemaItem),
+        memo,
+      )
+
+    if (schemaItem.items && Array.isArray(valueItem))
+      return R.assoc(
+        key,
+        valueItem.map((v) => scanSchemaAndPrefillValues(callback, v, schemaItem.items)),
+        memo,
+      )
+
+    const preDefinedValue = callback(schemaItem)
```

Comment:
it would be nice to rename `callback` into smth more descriptive, like `getValue` or smth